File: volume_1/chapter_5/langchain/llm_memory_persist_gateway.js
/**
 * LLM Memory History Gateway - LangChain JS with persistent session memory.
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

import { FileChatMessageHistory } from '@langchain/community/stores/message/file';
import { LangChainLLMManager as InMemoryLangChainLLMManager } from './llm_memory_gateway.js';
import { interactiveCli } from '../../chapter_4/utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

class LangChainLLMManager extends InMemoryLangChainLLMManager {
    constructor(memoryEnabled = true) {
        super(memoryEnabled);
        this.framework = 'LangChain+History JS';
    }

    _sessionFilePath(provider, sessionId) {
        const sessionsDir = path.join(__dirname, 'sessions');
        fs.mkdirSync(sessionsDir, { recursive: true });
        return path.join(sessionsDir, `${provider}__${sessionId}.json`);
    }

    _getHistory(provider, sessionId) {
        const key = this._historyKey(provider, sessionId);
        if (!this.histories.has(key)) {
            const filePath = this._sessionFilePath(provider, sessionId);
            this.histories.set(key, new FileChatMessageHistory({ filePath }));
        }
        return this.histories.get(key);
    }

    resetMemory(provider = null, sessionId = null) {
        const result = super.resetMemory(provider, sessionId);
        const sessionsDir = path.join(__dirname, 'sessions');

        if (result.removedSessions.length === 1 && result.removedSessions[0] === 'ALL') {
            if (fs.existsSync(sessionsDir)) {
                for (const file of fs.readdirSync(sessionsDir)) {
                    if (file.endsWith('.json')) {
                        fs.unlinkSync(path.join(sessionsDir, file));
                    }
                }
            }
            return result;
        }

        for (const key of result.removedSessions) {
            if (Array.isArray(key) && key.length === 2) {
                const [p, s] = key;
                const file = this._sessionFilePath(p, s);
                if (fs.existsSync(file)) {
                    fs.unlinkSync(file);
                }
            }
        }

        return result;
    }
}

async function main() {
    const args = process.argv.slice(2);
    if (args.length > 0 && args[0] === 'web') {
        const { runWebServer } = await import('../../chapter_4/web.js');
        await runWebServer(() => new LangChainLLMManager(true));
    } else {
        const manager = new LangChainLLMManager(true);
        await manager._checkProviders();
        await interactiveCli(manager);
    }
}

export { LangChainLLMManager };

if (import.meta.url === `file://${process.argv[1]}`) {
    main().catch(console.error);
}

File: volume_1/chapter_5/langchain/llm_memory_persist_gateway.py
"""LLM Memory History Gateway - LangChain with persistent session memory."""

import os
import sys
from pathlib import Path
from typing import Dict

CHAPTER_4_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", "chapter_4"))
sys.path.append(CHAPTER_4_ROOT)
sys.path.append(os.path.dirname(os.path.abspath(__file__)))

from langchain_community.chat_message_histories import FileChatMessageHistory

from llm_memory_gateway import LangChainLLMManager as InMemoryLangChainLLMManager
from utils import interactive_cli


class LangChainLLMManager(InMemoryLangChainLLMManager):
    """Persistent-memory variant that reuses the in-memory manager behavior."""

    def __init__(self, memory_enabled: bool = True):
        super().__init__(memory_enabled=memory_enabled)
        self.framework = "LangChain+History"

    def _session_file_path(self, provider: str, session_id: str) -> Path:
        sessions_dir = Path(__file__).resolve().parent / "sessions"
        sessions_dir.mkdir(parents=True, exist_ok=True)
        return sessions_dir / f"{provider}__{session_id}.json"

    def _get_history(self, provider: str, session_id: str) -> FileChatMessageHistory:
        key = (provider, session_id)
        if key not in self.histories:
            self.histories[key] = FileChatMessageHistory(file_path=str(self._session_file_path(provider, session_id)))
        return self.histories[key]

    def reset_memory(self, provider: str = None, session_id: str = None) -> Dict:
        result = super().reset_memory(provider=provider, session_id=session_id)

        if result["removed_sessions"] == ["ALL"]:
            for path in (Path(__file__).resolve().parent / "sessions").glob("*.json"):
                path.unlink(missing_ok=True)
            return result

        for key in result["removed_sessions"]:
            if isinstance(key, tuple):
                self._session_file_path(*key).unlink(missing_ok=True)

        return result


def main():
    if len(sys.argv) > 1 and sys.argv[1] == "web":
        from web import run_web_server

        run_web_server(lambda: LangChainLLMManager(memory_enabled=True))
    else:
        interactive_cli(LangChainLLMManager(memory_enabled=True))


if __name__ == "__main__":
    main()
